rfcs/rdfa-editor-stage-1.md

```diff
@@ -0,0 +1,587 @@
+
+##### Touching
+
+A range touches a node if it includes it, and if either the start is directly _after_ the closing tag of the node, _or_ the end is directly _before_ the opening tag.
+
+##### Inside
+
+A range is inside a node if the start lies _after_ the opening tag of the node, _and_ the end lies _before_ the closing tag.
+
+```typescript
+enum RangeContextStrategy {
+  CONTAINING, INCLUDING, TOUCHING, INSIDE
+}
+```
+
+
+#### ModelPosition (immutable)
+
+A ModelPosition represents a valid location in the document. 
+
+
+```typescript
+interface ModelPositon {
+  path: number[];
+  root: ModelNode;
+}
+```
+
+##### motivation
+
+If we take a step back, and ask ourselves what an intuitive definition for a Position in the document
+could be, we find ourselves torn between two answers.
+On the one hand, if we look at the rendered textDocument, it seems logical to represent
+a position as a single number, aka the number of characters from the start of the document.
+
+It is immediately clear why this is a useful representation. Arguably the most essential editing
+operation of them all, a single keypress of a character key, adds a single character to the document
+at the cursor position. all characters after it visually move by "1 spot". The same goes for the 
+behavior of an arrow key. We expect the cursor to move "1 step" to the next character. 
+
+However, once we get more familiar with the tree structure, problems start to appear. 
+A single "character offset" is simply not descriptive enough to specify a unique location in a tree.
+
+A simple example of this problem:
+
+```html
+<div>abc<span>def</span></div>
+```
+
+Here, while we can likely agree the unique position of character offsets 0,1 and 2, what position
+should an offset of 3 represent? After the "c", or before the "d"? There is no visual difference 
+of course, but anyone who has worked on this project knows that knowing the difference is vital
+for implementing the algorithms that actually do the work.
+
+Prosemirror proposes a simple solution to this problem: simply treat a node-boundary as a character.
+This would mean that the abovementioned 3rd position would be definitively set as "after c", and the 4th position
+would then be "before d".
+This does mean that we lose the ability to express "just move one character, regardless of tree structure". 
+
+Another way to represent positions is to embrace the tree structure and represent positions as paths of offsets from
+the root node. This is how the current ModelPositions work.
+The benefit of this representation is that it encodes way more information about the tree structure it operates in. It is
+also easier to calculate.
+
+As this second representation is currently implemented, this will be what we will use. However,
+the main reason for the small detour about character offsets above serves to illustrate that 
+both representations have merit, and a conversion between the two could be desirable.
```

Comment:
It would be very simple for the arrow handlers if we could just do position.getDocumentOffset() and ModelPosition.fromDocumentOffset() or something like that

Reply:
I think for arrow handlers I'd still like to lean on browser fallbacks
but yeah I think both systems can be useful. I don't think its a priority though